File: extapis/cgi/demoapps/shortest_path/shortest_path.cpp
#include <unistd.h>

#include "cgi.hpp"

#define INFINITY       1e+99

using namespace graphlab;
namespace json = rapidjson;

/**
 * Converts cstring to double. If cstring is empty, return INFINITY.
 */
static double stof(const char *str){
  if (0 == strlen(str)) return INFINITY;
  return atof(str);
}

/**
 * CGI Handler - contains methods (callbacks) to handle JSON invocations. Refer
 * to main GraphLab documentation for description of these methods.
 */
class cgi_handler : public icgi_handler {
public:
  
  void gather_edges(json::Document& invocation, json::Document& return_json){
    // we want to gather all incoming edges
    json::Document::AllocatorType& allocator = return_json.GetAllocator();
    return_json.AddMember("edges", "IN_EDGES", allocator);
  }
  
  void scatter_edges(json::Document& invocation, json::Document& return_json){
    // scatter to all outgoing edges
    json::Document::AllocatorType& allocator = return_json.GetAllocator();
    return_json.AddMember("edges", "OUT_EDGES", allocator);
  }
  
  void gather(json::Document& invocation, json::Document& return_json){
    json::Document::AllocatorType& allocator = return_json.GetAllocator();
    // find distance from source to this vertex thru neighbor
    const json::Value& edge = invocation["params"]["edge"];
    double edge_dist = atof(edge["state"].GetString());
    double nbr_dist = stof(edge["source"]["state"].GetString());
    double vertex_dist = nbr_dist + edge_dist;
    // return gather result
    std::ostringstream double_stream;
    double_stream << vertex_dist;
    const char *double_str = double_stream.str().c_str();
    json::Value dist(double_str, allocator);
    return_json.AddMember("result", dist, allocator);
  }
  
  void merge(json::Document& invocation, json::Document& return_json){
    json::Document::AllocatorType& allocator = return_json.GetAllocator();
    // keep the shortest distance
    double left = stof(invocation["state"].GetString());
    double right = stof(invocation["params"]["other"].GetString());
    double min_dist = std::min(left, right);
    // return merged gather result
    std::ostringstream double_stream;
    double_stream << min_dist;
    const char *double_str = double_stream.str().c_str();
    json::Value dist(double_str, allocator);
    return_json.AddMember("result", dist, allocator);
  }
  
  void apply(json::Document& invocation, json::Document& return_json){
    json::Document::AllocatorType& allocator = return_json.GetAllocator();
    // current vertex distance
    const char *vertex_state = invocation["params"]["vertex"]["state"].GetString();
    double vertex_dist = stof(vertex_state);
    // best distance found from gather
    double best_dist = stof(invocation["params"]["gather"].GetString());
    // if improvement, update
    if (best_dist < vertex_dist){
      std::ostringstream double_stream;
      double_stream << best_dist;
      const char *double_str = double_stream.str().c_str();
      json::Value dist(double_str, allocator);
      return_json.AddMember("vertex", dist, allocator);
    }
  }
  
  void scatter(json::Document& invocation, json::Document& return_json){
    json::Document::AllocatorType& allocator = return_json.GetAllocator();
    // best distance found for this vertex
    double best_dist = stof(invocation["params"]["vertex"]["state"].GetString());
    // parse distances
    const json::Value& edge = invocation["params"]["edge"];
    double nbr_dist = stof(edge["target"]["state"].GetString());
    double edge_dist = atof(edge["state"].GetString());
    // relax edge
    if (best_dist < INFINITY && nbr_dist > (best_dist + edge_dist))
      return_json.AddMember("signal", "TARGET", allocator);
  }
  
};

int main(int argc, char** argv) {

  cgi_handler handler;
  graphlab::cgi client(handler); 
  client.listen();
  return 0;

}

File: extapis/cgi/demoapps/shortest_path/shortest_path.rb
#!/usr/bin/env ruby

require 'json'

class String
  alias :old_to_f :to_f
  def to_f
    return 1e99 if 0 >= self.length
    self.old_to_f
  end
end

class Handler

  def gather_edges(json)
    {:edges => :IN_EDGES}
  end

  def scatter_edges(json)
    {:edges => :OUT_EDGES}
  end

  def gather(json)
    edge = json["params"]["edge"]
    # get distance to this vertex thru edge
    edge_dist = edge["state"].to_f
    nbr_dist = edge["source"]["state"].to_f
    vertex_dist = edge_dist + nbr_dist
    # return gather_type
    {:result => vertex_dist.to_s}
  end

  def merge(json)
    # take min of two gathers
    left = json["state"].to_f
    right = json["params"]["other"].to_f
    {:result => [left, right].min.to_s}
  end

  def apply(json)
    # take min of vertex dist and gather dist
    vertex_dist = json["params"]["vertex"]["state"].to_f
    gather_dist = json["params"]["gather"].to_f
    return {:vertex => gather_dist.to_s} if (gather_dist < vertex_dist)
    return {}
  end

  def scatter(json)
    edge = json["params"]["edge"]
    # get distance to neighbor thru this edge
    vertex_dist = json["params"]["vertex"]["state"].to_f
    edge_dist = edge["state"].to_f
    nbr_dist = edge["target"]["state"].to_f
    return {:signal => :TARGET} if (nbr_dist > (vertex_dist + edge_dist))
    return {}
  end

end

h = Handler.new

begin

  # first line tells us some many bytes
  bytes = STDIN.readline.to_i

  json = JSON.parse(STDIN.read bytes)
  method = json["method"]

  # invoke
  raise IOError, "Missing method field" if (nil == method)
  break if "exit" == method
  raise IOError, "Unrecognized method: " + method if not h.respond_to? method
  json = h.send method, json

  # return
  return_str = JSON.generate json
  puts return_str.length
  print return_str
  STDOUT.flush

rescue EOFError
  STDERR.write "Pipe broken"
end while true
